compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -30,6 +30,67 @@
 #include "il/Node_inlines.hpp"
 #include "infra/Bit.hpp"
 
+/**
+ * Generic helper that handles a number of similar binary operations.
+ * @param[in] node : calling node
+ * @param[in] regOp : the target AArch64 instruction opcode
+ * @param[in] regOp : the matching AArch64 immediate instruction opcode
+ * @param[in] cg : codegenerator
+ * @return target register
+ */
+static inline TR::Register *
+genericBinaryEvaluator(TR::Node *node, TR::InstOpCode::Mnemonic regOp, TR::InstOpCode::Mnemonic regOpImm, TR::CodeGenerator *cg)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Register *src1Reg = cg->evaluate(firstChild);
+   TR::Register *src2Reg = NULL;
+   TR::Register *trgReg = NULL;
+   bool tempRegUsed = false;
+   bool isImm = false;
+
+   if(1 == firstChild->getReferenceCount())
+      {
+      trgReg = src1Reg;
+      }
+   else if(1 == secondChild->getReferenceCount() && secondChild->getRegister() != NULL)
+      {
+      trgReg = src2Reg;
+      }
+   else
+      {
+      trgReg = cg->allocateRegister();
+      }
+
+   if (secondChild->getOpCode().isLoadConst() && secondChild->getRegister() == NULL)
+      {
+      int64_t value = secondChild->getLongInt();
+      /* When regOp = regOpImm, an immediate version of the instruction does not exist. */
+      if(constantIsUnsignedImm12(value) && regOp != regOpImm)
+         {
+         generateTrg1Src1ImmInstruction(cg, regOpImm, node, trgReg, src1Reg, value);
+         isImm = true;
+         }
+      else
+         {
+         src2Reg = cg->allocateRegister();
+         loadConstant64(cg, node, value, src2Reg);
```

Comment:
Because `loadConstant64` has the potential of generating multiple instructions, should you be distinguishing between 32-bit and 64-bit constants here?  i.e., if you're loading an integer constant, will you sometimes end up generating more instructions than you need because you are using `loadConstant64` rather than `loadConstant32` ?

Reply:
I think it may be worth it in the long run to check, yes.
@aarongraham9 @jeanlego @marufunb , any thoughts on this?

Reply:
Yes. I think checking for and distinguishing between 32-bit and 64-bit `const` loads would be a better idea here. Any further thoughts @jeanlego & @marufunb?